### ISSUE TITLE
GitHub Desktop forgets the GitHub Enterprise server URL

### ISSUE THREAD

**Issue by ittaibaratz** [desktop/desktop#12943]
### Describe the bug
Our GitHub Enterprise server is behind the company firewall and only accessible via VPN.

With the recent versions of GH Desktop (Since 2-3 weeks ago approx), every time I log off of VPN, GH desktop forgets my enterprise URL and I have to type it again.

While its expected that GH won't be able to communicate with the enterprise server when its not accessible, having to manually type in the URL every time I reconnect to VPN is tedious - And was not required with previous versions of the desktop tool.

### Version & OS

Version 2.9.3 (x64)

Mac OS Big Sur

### Steps to reproduce the behavior

1. Confirm GH Enterprise server is configured in preferences:

![image](https://user-images.githubusercontent.com/1454683/132885182-36eddecc-575c-42ca-9d36-2492490f98bf.png)

2. Disconnect from VPN so that GH Enterprise is not available to GH desktop.

3. Try to fetch origin, error message is shown (correctly) since GH Enterprise is not reachable.

4. You get an error that the account token is invalidated. This is actually not entirely accurate - Account token is still valid, just that the service is not reachable:

![image](https://user-images.githubusercontent.com/1454683/132885403-8824a0bc-cd37-4a0f-888a-1aa5f07ea83e.png)
 
5. Click "No" (Not on VPN now, no need to try again)

6. Connect to VPN again - GH Enterprise is reachable

7. Click "Fetch Origin" again - Error message shows up. Click "Open Preferences"

8. GH Enterprise is completely forgotten, need to type the URL again to use GH desktop:

![image](https://user-images.githubusercontent.com/1454683/132885689-c6df5cfe-0b45-4a0c-a09f-c3b31ba18ae2.png)

![image](https://user-images.githubusercontent.com/1454683/132885712-b53ebea5-a494-4146-b8bd-45463dd718c3.png)

### Expected behavior

I should be able to use GH Desktop with and without access to GH Enterprise for local development, and not have to re-type the Enterprise URL every time.

### Actual behavior

Need to provide the GH enterprise URL on every connection to VPN.

### Screenshots

Screenshots added above.

### Logs
[2021-09-10.desktop.production.log](https://github.com/desktop/desktop/files/7145152/2021-09-10.desktop.production.log)

Attach your logs by opening the `Help` menu and selecting `Show Logs...`, if applicable.

**Comment by steveward** [comment]
Thanks for the issue @ittaibaratz. It looks like the issue is your token is becoming invalid and returning a 401, which is triggering the authentication flow again:

```
2021-09-10T16:08:34.895Z - info: [ui] [AppStore] removing account ibaratz (Ittai Baratz) from store
2021-09-10T16:08:34.952Z - warn: [ui] deserialize: invalid JSON found at 'https://git.corp.adobe.com/api/v3/repos/exc/unified-shell/pulls?state=all&sort=updated&direction=desc&per_page=10' - status: 401, length: '33' id: '(missing)'
SyntaxError: Unexpected token P in JSON at position 0
```

This behavior is the result of improvements that were made in https://github.com/desktop/desktop/pull/12554 where the app now detects authentication errors and prompts you to authenticate again.

Is there anything in your network that could be interfering with the authentication request to your GitHub Enterprise server? 

We can also try investigating the request by using the Developer Tools. Here's how to do that:

 - launch the app, then click **View | Toggle Developer Tools**
 - switch to the **Network** tab
 - trigger the issue by attempting to log in
 - select the problem request and inspect the **Headers** and **Response**  to see if there is any useful information

Let me know if that shows anything interesting.

**Comment by ittaibaratz** [comment]
@steveward thanks a lot for the prompt response to this issue!

Looking at the developer tools, I can see this request returns 401:
`https://git.corp.adobe.com/api/v3/repos/<MY REPO AND ORG>/branches/help-center-react-intl/push_control`

The body of the response is:
```
Please refresh the page to login
```

I believe I have a certificate installed in the browser that allows access to GH Enterprise outside of the VPN, but GH desktop can not leverage it (Or I don't know how to make it leverage it), which is why I always use GH desktop with VPN when I need to push/pull from origin.

**Comment by steveward** [comment]
Thanks for checking that @ittaibaratz. I'm going to ping @sergiou87 on this to see if there's a way we can handle these types of scenarios since it's definitely painful to have to authenticate again each time you reconnect to a VPN.

**Comment by sergiou87** [comment]
Thank you both for looking into this! I will investigate if there is a nice way to tell apart these 401 and not prompt to re-login in that case 😄 

Sorry for all the trouble @ittaibaratz!!

**Comment by sergiou87** [comment]
@ittaibaratz could you please follow the steps @steveward mentioned above to look at the headers of the 401 response and see if you can find these starting with `x-github-`? I'm hoping you **cannot** see any of them 🤞 

![image](https://user-images.githubusercontent.com/1083228/133594692-778583c8-f9bd-4db5-8f94-cdbb621cf6d3.png)

**Comment by ittaibaratz** [comment]
@sergiou87 - Confirmed! No GH headers whatsoever. Looking forward to testing your fix. Thanks so much for getting this resolved so quickly.

**Comment by sergiou87** [comment]
Hey @ittaibaratz! The fix should be out now with version 2.9.4. Let me know if that helps 😄 

Thank you for your patience 🙇‍♂️ ❤️

**Comment by ittaibaratz** [comment]
Works like a charm! Thanks @sergiou87  and @steveward !
